runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/GrpcWindmillServer.java

```diff
@@ -1412,4 +1451,83 @@ public void close() throws IOException {
       stream.close();
     }
   }
+
+  /**
+   * A set of stopwatches used to track the amount of time spent throttled due to Resource Exhausted
+   * errors. Throttle time is cumulative for all three rpcs types but not for all streams. So if
+   * GetWork and CommitWork are both blocked for x, totalTime will be 2x. However, if 2 GetWork
+   * streams are both blocked for x totalTime will be x. All methods are thread safe.
+   */
+  private static class ThrottleTimer {
```

Comment:
Is this called from multiple threads or only from a single one?

Reply:
My understanding is that each Stream operates in its own thread owned by its executor.  As all streams share the throttle timer I believe it is being accessed by multiple threads regularly.

The executor is declared [here](https://github.com/apache/beam/pull/7664/files/762abacb2c91fb86e6ac7d2e1fbc874e832ee2a9#diff-b3c6a3d87f2bc409896e7e84a12f5d2dR574) at line 574
and gets used [here](https://github.com/apache/beam/pull/7664/files/762abacb2c91fb86e6ac7d2e1fbc874e832ee2a9#diff-b3c6a3d87f2bc409896e7e84a12f5d2dR717) at line 717